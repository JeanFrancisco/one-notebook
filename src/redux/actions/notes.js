import Swal from 'sweetalert2';
import { addDoc, collection, deleteDoc, doc, getDocs, orderBy, updateDoc } from '@firebase/firestore';
import {
    NOTES_NEW_NOTE,
    NOTES_SET_ACTIVE,
    NOTES_DISPLAY_IN_SIDEBAR,
    NOTES_UPDATED_NOTE,
    NOTES_DROP_NOTE,
    NOTES_CLEANING_WHEN_LOGOUT,
    NOTES_CHANGE_ORDER,
} from '../constants/typesNotes';
import { uploadFile } from '../../helpers/files';
import { db as firestoreDB } from '../../config/firebase';

export const actionAddNewNote = () => {
    return async (dispatch, getState) => {
        const { uid: user_uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            urlImage: '',
            createdAt: new Date().getTime(),
        };

        const doc = await addDoc( collection(firestoreDB, `${user_uid}/journal/notes`), newNote);

        const createdNote = {
            id: doc.id,
            ...newNote,
        };

        dispatch( actionNewNote( createdNote ) );
        dispatch( actionSetActiveNote( createdNote ) );
    }
};

export const actionNewNote = ( createdNote ) => ({
    type: NOTES_NEW_NOTE,
    payload: createdNote,
});

export const actionSetActiveNote = ( activeNote ) => ({
    type: NOTES_SET_ACTIVE,
    payload: activeNote,
});

export const actionSavingChanges = () => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;
        const activeNote = getState().notes.active;

        const { id: note_id, ...noteForFirestore } = activeNote;
        await updateDoc( doc(firestoreDB, `${user_uid}/journal/notes/${ note_id }`), noteForFirestore )
            .then(() => {
                dispatch( refreshNote(activeNote) );

                Swal.fire({
                    title: 'Guardada exitosamente',
                    html: '<p><small>Su nota se guardó correctamente</small></p>',
                    icon: 'success',
                    timer: 1300,
                    showConfirmButton: true,
                    confirmButtonText: 'Aceptar',
                    confirmButtonColor: 'limegreen',
                });
            })
            .catch(reason => {
                console.log( reason.code );
            });
    }
}

export const refreshNote = ( updatedNote ) => ({
    type: NOTES_UPDATED_NOTE,
    payload: { ...updatedNote },
});

export const getFirebaseNotes = () => {
    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        const snapNotes = await getDocs( collection(firestoreDB, `${uid}/journal/notes`) );
        const notes = [];

        snapNotes.forEach( snapHijo => {
            notes.push({
                id: snapHijo.id,
                ...snapHijo.data(),
            })
        });

        dispatch( displayNotes(notes) );
    }
}

export const displayNotes = ( notes ) => ({
    type: NOTES_DISPLAY_IN_SIDEBAR,
    payload: notes,
});

export const actionSetAscendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'asc',
    }
});

export const actionSetDescendingOrder = () => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        order: 'desc',
    }
});

export const actionReorderByRule = ( orderBy ) => ({
    type: NOTES_CHANGE_ORDER,
    payload: {
        orderby: orderBy,
    }
});

export const actionStartUploadFile = (file) => {
    return async (dispatch, getState) => {
 
        Swal.fire({
            title: 'Subiendo archivos...',
            text: 'Por favor, espere un momento...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            },
        });

        const activeNote = getState().notes.active;
        const { secure_url } = await uploadFile(file);

        dispatch( actionSetActiveNote({ ...activeNote, urlImage: secure_url }) );
        await dispatch( actionSavingChanges() );

        Swal.close();
    }
}

export const actionStartDeletingNote = (note_id) => {
    return async (dispatch, getState) => {
        const user_uid = getState().auth.uid;

        try {
            const resp = await deleteDoc( doc( firestoreDB, `${user_uid}/journal/notes/${note_id}`) );

            dispatch( deleteNote( note_id ) );

            Swal.fire({
                icon: 'success',
                timer: 1250,
                title: 'Eliminado',
                text: 'Su nota fue eliminada',
                confirmButtonText: 'Aceptar',
                confirmButtonColor: 'green',
                showCloseButton: true,
            });

        } catch (e) {
            Swal.fire({
                icon: 'error',
                html: 'Error en la conexión con Firestore DB',
                title: 'Error. <small>No pudimos terminar la eliminación</small>',
                footer: `<small>For more details: ${e.code}</small>`,
                confirmButtonText: 'Entendido',
                confirmButtonColor: 'oragered', 
                showCloseButton: true,
            });

            console.error(e);
        }
    }
}

export const deleteNote = (noteID) => ({
    type: NOTES_DROP_NOTE,
    payload: noteID
});

export const cleanupNotesLogout = () => ({
    type: NOTES_CLEANING_WHEN_LOGOUT
});