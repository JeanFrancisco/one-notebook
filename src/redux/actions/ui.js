import { UI_FINISH_LOADING, UI_START_LOADING } from "../constants/typesUi";

export const actionStartLoading = () => ({
    type: UI_START_LOADING,
});

export const actionFinishLoading = () => ({
    type: UI_FINISH_LOADING,
});