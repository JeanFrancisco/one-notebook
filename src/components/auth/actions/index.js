import {
    createUserWithEmailAndPassword,
    signInWithEmailAndPassword,
    GoogleAuthProvider,
    signInWithPopup,
    updateProfile,
    getAuth,
    signOut,
} from "firebase/auth";
import Swal from 'sweetalert2';
import { app } from "../../../config/firebase";
import { cleanupNotesLogout } from "../../../redux/actions/notes";
import { actionFinishLoading, actionStartLoading } from "../../../redux/actions/ui";
import {
    AUTH_LOGIN,
    AUTH_LOGOUT,
    AUTH_SET_ERROR,
    AUTH_REMOVE_ERROR
} from "../constants";

export const authLogin = (uid, displayName) => ({
    type: AUTH_LOGIN,
    payload: {
        uid,
        displayName,
    }
});

export const authSignInWithGoogle = () => {
    return async (dispatch) => {

        dispatch( actionStartLoading() );

        const auth = getAuth( app );

        await signInWithPopup( auth, new GoogleAuthProvider() )
        .then( userCredentials => {
            // const { user: { displayName, uid } } = userCredentials;
            
            // dispatch( authLogin(uid, displayName) );
        })
        .catch( reason => {
            Swal.fire({
                icon: 'warning',
                html: '<small>No se pudo iniciar la sesión. <br/> <br/> ¿Alcanzó a ver si ocurrió algo: con el Pop-up de autenticación de Google?</small>',
                title: 'Oups! Lo sentimos.',
                iconHtml: '<i class="fa fa-warning"></i>',
                footer: `<small>Reason Code: ${reason.code}</small>`,
                showCloseButton: true,
                confirmButtonText: 'Aceptar',
            });
        })
        .finally(() => {
            dispatch( actionFinishLoading() );
        });
    };
}

export const authEmailPasswordSignIn = ({ email, password }) => {
    return async dispatch => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await signInWithEmailAndPassword(auth, email, password)
            .then( ({ user }) => {
                // const { uid, displayName } = user;

                // dispatch( authLogin(uid, displayName) );
            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Algo ocurrió mientras iniciabamos tu sesión.</small>',
                    html: `<p><b><small>¿Qué acciones puedes hacer?</small></b></p>
                        <ul style="text-align: left">
                            <li><small>Puedes revisar por favor tu email y contraseña.</small></li>
                            <li><small>También que te puedes conectar a Internet. En especial a los servicios de Google.</small></li>
                        </ul>`,
                    confirmButtonText: 'Entendido',
                    icon: 'warning',
                    iconColor: 'orange',
                    showCloseButton: true,
                    confirmButtonColor: 'orangered',
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                })
            }).finally(() => {
                dispatch( actionFinishLoading() );
            });
    }
}

export const authEmailPasswordSignUp = ({ email, password, username }) => {
    return async (dispatch) => {
        dispatch( actionStartLoading() );

        const auth = getAuth(app);

        await createUserWithEmailAndPassword(auth, email, password)
            .then( async userCredentials => {

                const { user } = userCredentials;
                await updateProfile(user, { displayName: username });

                // dispatch( authLogin(user.uid, user.displayName) );

            }).catch( reason => {
                Swal.fire({
                    title: 'Oups! <small>Ahora mismo no podemos registrarte.</small>',
                    html: '<p>Por favor... <br/> <small>Regresa más tarde y vuelve a intentarlo ;) .</small><p>',
                    icon: 'info',
                    iconColor: 'orange',
                    confirmButtonText: 'Entiendo',
                    confirmButtonColor: 'limegreen',
                    showCloseButton: true,
                    footer: `<small>Reason Code: ${reason.code}</small>`,
                });
            }).finally(() => {
               dispatch( actionFinishLoading() );
            });
    }
}

export const authFirebaseLogout = () => {
    return async dispatch => {
        const auth = getAuth(app);

        await signOut(auth);

        dispatch( authLogout() );

        dispatch( cleanupNotesLogout() );
    }
}

export const authLogout = () => ({ type: AUTH_LOGOUT });

export const setAuthError = ( msg ) => ({
    type: AUTH_SET_ERROR,
    payload: msg
});

export const removeAuthError = () => ({ type: AUTH_REMOVE_ERROR });