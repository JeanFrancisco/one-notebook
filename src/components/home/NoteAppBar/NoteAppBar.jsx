import { useRef } from 'react';
import { useSelector } from 'react-redux';
import { useDispatch } from 'react-redux';
import localMoment from '../../../config/moment';
import { actionSavingChanges, actionStartUploadFile } from '../../../redux/actions/notes';

const NoteAppBar = () => {
    const dispatch = useDispatch();
    const activeNote = useSelector( store => store.notes.active );

    const inputFile = useRef();

    const handleSavingNote = () => {
        dispatch( actionSavingChanges() );
    }

    const handleUploadClick = () => {
        // document.getElementById('selector__file-photo').click(); // Trigger the click event listener..
        inputFile.current.click();
    }

    const handleChangeSelectFile = e => {
        const uniqueFile = e.target.files[0];

        dispatch( actionStartUploadFile(uniqueFile) );
    }

    return <div className="note__appbar">
        <span className="note__appbar-date">{ localMoment(activeNote?.createdAt).format('LLLA') }</span>
        <input id="selector__file-photo"
            type='file'
            style={{ display: 'none' }}
            onChange={ handleChangeSelectFile }
            ref={ inputFile }
            />
        <button className="btn btn-primary"
            onClick={ handleUploadClick } >
            <span className="fas fa-cloud-upload-alt fa-sm"> </span>
            &nbsp;
            Subir Imagen
        </button>
        <button className="btn btn-primary"
            onClick={ handleSavingNote }
            >
            <i className="fas far fa-save fa-sm"></i>
            &nbsp;
            Guardar
        </button>
    </div>;
};

export default NoteAppBar;