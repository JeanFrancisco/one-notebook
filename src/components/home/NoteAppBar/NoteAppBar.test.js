import React from 'react';
import { shallow } from 'enzyme';
import { actionSavingChanges, actionStartUploadFile } from '../../../redux/actions/notes';
import NoteAppBar from './NoteAppBar';

const clickInputFile = jest.fn();

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
    useDispatch: jest.fn( () => mockDispatch ),
}));

jest.mock('../../../redux/actions/notes', () => ({
    actionSavingChanges: jest.fn(),
    actionStartUploadFile: jest.fn(),
}));

describe('Pruebas al componente de <NoteAppBar/>.jsx', () => {

    const wrapper = shallow(<NoteAppBar />);

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('Debe hacer el match con el snapshot capturado por defecto', () => {
        expect( wrapper ).toMatchSnapshot();
    });


    /* Trying to modify the onClick function, but it fails.
     * Due the setProps method only can be called at the root element */

    // test('Debe ejecutar el click del input con type="file", cuando dé click en el botón de Subir Imagen', () => {
    //     const handleClickFileInput = jest.fn();

    //     wrapper.find('#selector__file-photo').setProps( 'onClick', handleClickFileInput );

    //     wrapper.find('button').first().simulate('click');

    //     expect( handleClickFileInput ).toHaveBeenCalled(1);
    // });

    /* Correct implementation using useRef hook */
    test('Debe ejecutar el click del input con type="file" (Referenciado con useRef), cuando dé click en el botón de Subir Imagen', () => {
        const useRefSpy = jest.spyOn(React, 'useRef').mockReturnValueOnce({ current: { click: clickInputFile } });

        const wrapper = shallow(<NoteAppBar />);

        wrapper.find('button').first().simulate('click');

        expect( clickInputFile ).toHaveBeenCalledTimes(1);
    });

    test('Debe ejecutar la acción actionSavingChanges(), cuando dé click en el botón de Guardar', () => {
        wrapper.find('button').last().simulate('click');

        expect( actionSavingChanges ).toHaveBeenCalledTimes(1);
    });

    test('Debe ejecutar la acción actionStartUploadFile(), cuando haya hecho cambios en el input de type="file"', () => {        
        const fakeFileTest = new File([], 'test_file.txt');

        wrapper.find('input[type="file"]').simulate('change', {
            target: {
                files: [
                    fakeFileTest,
                ],
            }
        });

        expect( actionStartUploadFile ).toHaveBeenCalledWith( fakeFileTest );
    });
});