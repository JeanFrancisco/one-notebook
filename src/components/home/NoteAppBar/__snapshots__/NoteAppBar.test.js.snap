// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente de <NoteAppBar/>.jsx Debe hacer el match con el snapshot capturado por defecto 1`] = `
<div
  className="note__appbar"
>
  <span
    className="note__appbar-date"
  >
    5 de febrero de 2022 13:47PM
  </span>
  <input
    id="selector__file-photo"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    <span
      className="fas fa-cloud-upload-alt fa-sm"
    >
       
    </span>
      Subir Imagen
  </button>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    <i
      className="fas far fa-save fa-sm"
    />
      Guardar
  </button>
</div>
`;
